inductor cpp wrapper: add GIL release back

chunyuan-w opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #123897

Fixes https://github.com/pytorch/pytorch/issues/123517.
This PR adds the GIL release (originally added in https://github.com/pytorch/pytorch/pull/111888) back following the suggestion here: https://github.com/pytorch/pytorch/pull/123897#discussion_r1562509705.
We added a default constructor and an assignment operator for the `RAIIPyObject` class
 (https://github.com/pytorch/pytorch/pull/123897#discussion_r1566262575) in order to declare the `custom_op_wrapper` outside of the GIL acquisition scope.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler @amjames @desertfire @chauhang

chunyuan-w committed:
inductor cpp wrapper: add GIL release back

[ghstack-poisoned]

chunyuan-w committed:
Update on "inductor cpp wrapper: add GIL release back"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

jgong5 commented on torch/_inductor/codegen/cpp_wrapper_cpu.py:
```diff
@@ -522,6 +520,8 @@ def write_wrapper_decl(self):
                             """
                         )
                     else:
+                        # release GIL to support multiple instances inference (in different threads of the same process)
+                        self.prefix.splice("pybind11::gil_scoped_release release;")                          
```
What's the story for `config.abi_compatible=True`?

chunyuan-w replied:
This line of GIL release has been removed in https://github.com/pytorch/pytorch/pull/122554.
I'm still checking if it's intended but it does cause regression in https://github.com/pytorch/pytorch/issues/123517.

For `abi_compatible`, I met some failures for UTs on CUDA after adding this line back. Still checking what happens.

desertfire replied:
Because in the ABI-compatible mode, we aren't allowed to call dispatcher Boxed call, instead we go back to Python to call custom ops in https://github.com/pytorch/pytorch/pull/122554/ (Note: this is only for the JIT mode). If this is really performance critical, you could gil_scoped_release by default, gil_scoped_acquire before entering the custom op call and destroy that gil_scoped_acquire after done with the custom op call.

chunyuan-w committed:
Update on "inductor cpp wrapper: add GIL release back"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

chunyuan-w committed:
Update on "inductor cpp wrapper: add GIL release back"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

chunyuan-w committed:
Update on "inductor cpp wrapper: add GIL release back"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

chunyuan-w committed:
Update on "inductor cpp wrapper: add GIL release back"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

peterbell10 commented on torch/_inductor/codegen/cpp_wrapper_cpu.py:
```diff
@@ -2011,11 +2025,20 @@ def load_custom_op_wrapper(self):
+
+        # Use ptr here because RAIIPyObject does not have a default constructor
+        declarations_before_scope = [
+            "std::unique_ptr<RAIIPyObject> custom_op_wrapper_ptr;"
```
Why not add a default constructor and assignment operator? Allocating a pointer in the heap and only using it locally seems wasteful.

chunyuan-w committed:
Update on "inductor cpp wrapper: add GIL release back"


Fixes https://github.com/pytorch/pytorch/issues/123517.
This PR adds the GIL release (originally added in https://github.com/pytorch/pytorch/pull/111888) back.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

chunyuan-w committed:
Update on "inductor cpp wrapper: add GIL release back"


Fixes https://github.com/pytorch/pytorch/issues/123517.
This PR adds the GIL release (originally added in https://github.com/pytorch/pytorch/pull/111888) back.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

chunyuan-w commented:
@pytorchbot merge
